table/block_based/block_based_table_builder.cc

```diff
@@ -788,6 +1012,61 @@ void BlockBasedTableBuilder::WriteRawBlock(const Slice& block_contents,
   }
 }
 
+void BlockBasedTableBuilder::BGWorkWriteRawBlock() {
+  Rep* r = rep_;
+  ParallelCompressionRep::BlockRepSlot* slot;
+  ParallelCompressionRep::BlockRep* block_rep;
+  while (r->pc_rep->write_queue.pop(slot)) {
```

Comment:
I think if r->status is OK, we should exit the loop and avoid calling `WriteRawBlock()` again. I believe inside `WriteRawBlock()` we only assert status OK and in release mode we would just override the status, which can cause problem. I think it's safer to just exit the loop if r->status.

Ideally the failure case is tested.

Sorry for this late comment. I should have noticed it earlier. But this is serious and we have to fix it before commit the feature.

Reply:
Sure. Did you mean exiting the loop when r->status is NOT OK?

Reply:
Yes. Sorry for the typo.

Reply:
Also, there seems to be a data race between the main thread and the block writer thread with rep_->status().

Those variables added to estimate file size might also have a data race.

Reply:
For variables to estimate file size,

`raw_bytes_submitted` will only be updated and accessed in block building thread, should be safe;

`raw_bytes_compressed` and `curr_compression_ratio` will be updated in writer thread, and accessed in block building thread. Their updates should be single-threaded. There might be the case where `r->offset` is already updated, but `raw_bytes_compressed` and `curr_compression_ratio` are not updated yet, causing an estimation error of 1-compressed-block-size. Shall we make these updates atomic, or you think it's acceptable?

Reply:
I noticed that, in orignal BlockBasedTableBuilder, when compression is aborted, `rep_->status` is set to `Status::Corruption`, but `WriteRawBlock` is still called. Is `WriteRawBlock` meant to write uncompressed data here, or we should return before `WriteRawBlock` is called?

Reply:
Regarding the data race for variables related to file size estimation, making the variables used by `BlockBasedTableBuilder::EstimatedFileSize` atomic should be good enough to me.

The compression validation failure seems to be a bug. We don't have to fix the bug with non-parallel case. But if it is fixed in parallel writer case, it is great.

Reply:
Variables used by `BlockBasedTableBuilder::EstimatedFileSize` is all protected by `estimation_mutex`. This should lead to more accurate (and predictable) estimation then separate atomic variables.

I've added `if (!ok())` check for compression in both parallel and non-parallel cases. A fake faulting compressor is needed in future for thorough testing though.